File: llm/schemas.py
"""
Pydantic schemas for Unified Generation Pipeline.
Strict JSON schemas ensure LLM outputs are validated and typed.
"""
from typing import Optional, List, Literal, Dict
from uuid import UUID
from pydantic import BaseModel, Field
from server.enums import (
    ConversationStage,
    IntentLevel,
    UserSentiment,
    DecisionAction,
    RiskLevel,
)


# ============================================================
# Pipeline Input Context
# ============================================================

class MessageContext(BaseModel):
    """A single message in conversation history."""
    sender: Literal["lead", "bot", "human"]
    text: str
    timestamp: str  # ISO format


class TimingContext(BaseModel):
    """Timing information for decisions."""
    now_local: str
    last_user_message_at: Optional[str] = None
    last_bot_message_at: Optional[str] = None
    whatsapp_window_open: bool = True


class NudgeContext(BaseModel):
    """Anti-spam tracking."""
    followup_count_24h: int = 0
    total_nudges: int = 0


class PipelineInput(BaseModel):
    """
    Complete input context for the pipeline.
    Kept minimal for token efficiency.
    """
    # Business context
    organization_id: UUID
    business_name: str
    business_description: str = ""
    flow_prompt: str = ""  # Conversation flow/sales script instructions
    
    # CTAs
    available_ctas: List[Dict[str, str]] = []  # [{id: UUID, name: str}]
    
    # Conversation context
    rolling_summary: str = ""
    last_messages: List[MessageContext] = []
    
    # Current state
    conversation_stage: ConversationStage
    conversation_mode: Literal["bot", "human"]
    intent_level: IntentLevel
    user_sentiment: UserSentiment
    active_cta_id: Optional[UUID] = None
    
    # Timing
    timing: TimingContext
    nudges: NudgeContext
    
    # Constraints (from business config)
    max_words: int = 80
    questions_per_message: int = 1
    language_pref: str = "en"
    
    # RAG Context (injected by Pipeline)
    dynamic_knowledge_context: Optional[str] = None


# ============================================================
# Unified Generation Step
# ============================================================

class RiskFlags(BaseModel):
    """Risk assessment for the conversation."""
    spam_risk: RiskLevel = RiskLevel.LOW
    policy_risk: RiskLevel = RiskLevel.LOW
    hallucination_risk: RiskLevel = RiskLevel.LOW


class GenerateOutput(BaseModel):
    """
    Unified output from the single LLM generation step.
    Combines reasoning, decision making, and response generation.
    """
    # Reasoning
    thought_process: str = Field(..., description="Chain of thought reasoning")
    
    # Observation / Classification
    intent_level: IntentLevel
    user_sentiment: UserSentiment
    risk_flags: RiskFlags
    
    # Decision
    action: DecisionAction
    new_stage: ConversationStage
    should_respond: bool = False
    
    # Action Payload
    selected_cta_id: Optional[UUID] = None
    cta_scheduled_at: Optional[str] = None
    followup_in_minutes: int = 0
    followup_reason: str = ""
    
    # Response
    message_text: str = ""
    message_language: str = "en"
    
    # Metadata
    confidence: float = Field(..., ge=0.0, le=1.0)
    needs_human_attention: bool = False


# ============================================================
# Step 4: Memory Output (Archivist)
# ============================================================

class MemoryOutput(BaseModel):
    """
    Output from Step 4: Memory (Archivist).
    Updated rolling summary.
    """
    updated_rolling_summary: str = Field(..., max_length=2000)
    needs_recursive_summary: bool = False


# Backward compatibility alias
SummaryOutput = MemoryOutput


# ============================================================
# Complete Pipeline Result
# ============================================================

class PipelineResult(BaseModel):
    """
    Complete result from running the pipeline.
    """
    # Unified output
    generate: GenerateOutput
    
    # Memory output (separate step)
    memory: Optional[MemoryOutput] = None
    
    # Metadata
    pipeline_latency_ms: int = 0
    total_tokens_used: int = 0
    
    # Async Flags
    needs_background_summary: bool = True
    
    # Computed action helpers
    @property
    def should_send_message(self) -> bool:
        return self.generate.should_respond and bool(self.generate.message_text)
    
    @property
    def should_schedule_followup(self) -> bool:
        return self.generate.action == DecisionAction.WAIT_SCHEDULE
    
    @property
    def should_escalate(self) -> bool:
        return self.generate.needs_human_attention
        
    @property
    def should_initiate_cta(self) -> bool:
        return self.generate.action == DecisionAction.INITIATE_CTA

File: llm-go/schemas/schemas.go
package schemas

import (
	"whatsapp-funnel/llm-go/enums"
)

type MessageContext struct {
	Sender    string `json:"sender"`
	Text      string `json:"text"`
	Timestamp string `json:"timestamp"`
}

type TimingContext struct {
	NowLocal           string  `json:"now_local"`
	LastUserMessageAt  *string `json:"last_user_message_at,omitempty"`
	LastBotMessageAt   *string `json:"last_bot_message_at,omitempty"`
	WhatsAppWindowOpen bool    `json:"whatsapp_window_open"`
}

type NudgeContext struct {
	FollowupCount24h int `json:"followup_count_24h"`
	TotalNudges      int `json:"total_nudges"`
}

type CTA struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type PipelineInput struct {
	OrganizationID          string                  `json:"organization_id"`
	BusinessName            string                  `json:"business_name"`
	BusinessDescription     string                  `json:"business_description"`
	FlowPrompt              string                  `json:"flow_prompt"`
	AvailableCTAs           []CTA                   `json:"available_ctas"`
	RollingSummary          string                  `json:"rolling_summary"`
	LastMessages            []MessageContext        `json:"last_messages"`
	ConversationStage       enums.ConversationStage `json:"conversation_stage"`
	ConversationMode        string                  `json:"conversation_mode"`
	IntentLevel             enums.IntentLevel       `json:"intent_level"`
	UserSentiment           enums.UserSentiment     `json:"user_sentiment"`
	ActiveCTAID             *string                 `json:"active_cta_id,omitempty"`
	Timing                  TimingContext           `json:"timing"`
	Nudges                  NudgeContext            `json:"nudges"`
	MaxWords                int                     `json:"max_words"`
	QuestionsPerMessage     int                     `json:"questions_per_message"`
	LanguagePref            string                  `json:"language_pref"`
	DynamicKnowledgeContext *string                 `json:"dynamic_knowledge_context,omitempty"`
}

type RiskFlags struct {
	SpamRisk          enums.RiskLevel `json:"spam_risk"`
	PolicyRisk        enums.RiskLevel `json:"policy_risk"`
	HallucinationRisk enums.RiskLevel `json:"hallucination_risk"`
}

type GenerateOutput struct {
	ThoughtProcess      string                  `json:"thought_process"`
	IntentLevel         enums.IntentLevel       `json:"intent_level"`
	UserSentiment       enums.UserSentiment     `json:"user_sentiment"`
	RiskFlags           RiskFlags               `json:"risk_flags"`
	Action              enums.DecisionAction    `json:"action"`
	NewStage            enums.ConversationStage `json:"new_stage"`
	ShouldRespond       bool                    `json:"should_respond"`
	SelectedCTAID       *string                 `json:"selected_cta_id"`
	CTAScheduledAt      *string                 `json:"cta_scheduled_at"`
	FollowupInMinutes   int                     `json:"followup_in_minutes"`
	FollowupReason      string                  `json:"followup_reason"`
	MessageText         string                  `json:"message_text"`
	MessageLanguage     string                  `json:"message_language"`
	Confidence          float64                 `json:"confidence"`
	NeedsHumanAttention bool                    `json:"needs_human_attention"`
}

type MemoryOutput struct {
	UpdatedRollingSummary string `json:"updated_rolling_summary"`
	NeedsRecursiveSummary bool   `json:"needs_recursive_summary"`
}

type PipelineResult struct {
	Generate               GenerateOutput `json:"generate"`
	Memory                 *MemoryOutput  `json:"memory,omitempty"`
	PipelineLatencyMs      int            `json:"pipeline_latency_ms"`
	TotalTokensUsed        int            `json:"total_tokens_used"`
	NeedsBackgroundSummary bool           `json:"needs_background_summary"`
}

func (p PipelineResult) ShouldSendMessage() bool {
	return p.Generate.ShouldRespond && p.Generate.MessageText != ""
}

func (p PipelineResult) ShouldScheduleFollowup() bool {
	return p.Generate.Action == enums.ActionWaitSchedule
}

func (p PipelineResult) ShouldEscalate() bool {
	return p.Generate.NeedsHumanAttention
}

func (p PipelineResult) ShouldInitiateCTA() bool {
	return p.Generate.Action == enums.ActionInitiateCTA
}
